File: embedded interview prep/Cpp/ContainsDuplicate.cpp
/**
 * @file ContainsDuplicate.cpp
 * @brief Implementation of duplicate detection in integer arrays
 * @author Mervin Nguyen
 * @date 2026-03-28
 * @version 1.0.0
 * 
 * Approach:
 * 1. Sort the array in order.
 * 2. Iterate through the array, comparing each element to the previous one.
 * 3. If any two consecutive elements are equal, return true (duplicate found).
 * 4. If no duplicates are found, return false.
 */

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std; 

class Solution {
public:
    /**
     * @brief Checks if an array contains any duplicate values
     * @param[in,out] nums Vector of integers to check (modified by sorting)
     * @return true if duplicates exist, false otherwise
     * @note Time complexity: O(n log n) due to sorting, Space complexity: O(1)
     */
    bool hasDuplicate(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for (int i = 1; i < nums.size(); i++){
            if (nums[i] == nums[i - 1]){
                return true;
            }
        }
        return false;
    }
};

int main(){
    Solution sol;
    vector<int> nums = {1, 4, 3, 2};
    bool result = sol.hasDuplicate(nums);
    cout << (result ? "true" : "false") << endl;
    return 0;
}

File: python solutions/ContainsDuplicate.py
"""Contains Duplicate

Given an integer array nums, return true if any value appears more than once 
in the array, otherwise return false.
"""

from typing import List


class Solution:
    """Solution class for the Contains Duplicate problem."""
    
    def hasDuplicate(self, nums: List[int]) -> bool:
        """Check if any value appears more than once in the array.
        
        Uses a hash map to track seen values while iterating through the array.
        Returns True as soon as a duplicate is found for O(1) early termination.
        
        Args:
            nums: List of integers to check for duplicates.
            
        Returns:
            True if any value appears more than once, False otherwise.
            
        Example:
            >>> solution = Solution()
            >>> solution.hasDuplicate([1, 2, 3, 1])
            True
            >>> solution.hasDuplicate([1, 2, 3, 4])
            False
        """
        hashmap = {}
        for i in nums:
            if i in hashmap:
                return True
            hashmap[i] = 1
        return False